# Retail loans classification

## 1. Import libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

## 2. Working with data

## 2.1. Import data

In [2]:
github_p = "https://raw.githubusercontent.com/Finance-781/FinML/master/Lecture%203%20-%20Classification/Inclass/"
df = pd.read_csv(github_p+'data/loans.csv')
print("Done")

Done


In [3]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

   int_rate  installment grade sub_grade  ... total_bal_il il_util  \
0     10.65       162.87     B        B2  ...          NaN     NaN   
1     15.27        59.83     C        C4  ...          NaN     NaN   
2     15.96        84.33     C        C5  ...          NaN     NaN   
3     13.49       339.31     C        C1  ...          NaN     NaN   
4     12.69        67.79     B        B5  ...          NaN     NaN   

  open_rv_12m  open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN          NaN        NaN      NaN              NaN    NaN   
1         NaN          NaN        NaN      NaN              NaN    NaN   
2         NaN          NaN        NaN      NaN              NaN    NaN   
3         NaN          NaN        NaN      NaN              NaN    NaN   
4         NaN          NaN        NaN      NaN              NaN    NaN   

  total_cu_tl inq_last_12m  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN  

[5 rows x 74 columns]

## 2.2. Data cleaning

### 2.2.1. Working with missing values

In [4]:
df.shape

(102388, 74)

In [5]:
df.isnull().sum()

id                                  0
member_id                           0
loan_amnt                           0
funded_amnt                         0
funded_amnt_inv                     0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_title                        6525
emp_length                       3875
home_ownership                      0
annual_inc                          4
verification_status                 0
issue_d                             0
loan_status                         0
pymnt_plan                          0
url                                 0
desc                            54706
purpose                             0
title                              15
zip_code                            0
addr_state                          0
dti                                 0
delinq_2yrs                        29
earliest_cr_

There are some columns with all null values, let's drop them

In [6]:
null_cols = [i for i in df.isnull().sum()== df.shape[0]]
df.drop(df.columns[null_cols],axis=1,inplace=True)

In [7]:
df.shape

(102388, 57)

In [8]:
df.isnull().sum()

id                                 0
member_id                          0
loan_amnt                          0
funded_amnt                        0
funded_amnt_inv                    0
term                               0
int_rate                           0
installment                        0
grade                              0
sub_grade                          0
emp_title                       6525
emp_length                      3875
home_ownership                     0
annual_inc                         4
verification_status                0
issue_d                            0
loan_status                        0
pymnt_plan                         0
url                                0
desc                           54706
purpose                            0
title                             15
zip_code                           0
addr_state                         0
dti                                0
delinq_2yrs                       29
earliest_cr_line                  29
i

In [9]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

   int_rate  installment grade sub_grade  ... next_pymnt_d last_credit_pull_d  \
0     10.65       162.87     B        B2  ...          NaN            1/01/16   
1     15.27        59.83     C        C4  ...          NaN            1/09/13   
2     15.96        84.33     C        C5  ...          NaN            1/01/16   
3     13.49       339.31     C        C1  ...          NaN            1/01/15   
4     12.69        67.79     B        B5  ...      1/02/16            1/01/16   

  collections_12_mths_ex_med  mths_since_last_major_derog policy_code  \
0                        0.0                          NaN           1   
1                        0.0                          NaN           1   
2                        0.0                          NaN           1   
3                        0.0                          NaN           1   
4                        0.0                          NaN           1   

  application_type acc_now_delinq tot_coll_amt tot_cur_bal total_rev_hi_lim  
0       INDIVIDUAL            0.0          NaN         NaN              NaN  
1       INDIVIDUAL            0.0          NaN         NaN              NaN  
2       INDIVIDUAL            0.0          NaN         NaN              NaN  
3       INDIVIDUAL            0.0          NaN         NaN              NaN  
4       INDIVIDUAL            0.0          NaN         NaN              NaN  

[5 rows x 57 columns]

Let's drop some columns which will not be used for futher analysis:

In [10]:
df.drop(['id', 'member_id', 'emp_title', 'url', 'desc', 'zip_code', 'title', 'addr_state'], axis=1, inplace=True)

In [11]:
df.shape

(102388, 49)

Let's see the percentage of missing values:

In [12]:
df.isnull().sum()/df.shape[0]*100

loan_amnt                       0.000000
funded_amnt                     0.000000
funded_amnt_inv                 0.000000
term                            0.000000
int_rate                        0.000000
installment                     0.000000
grade                           0.000000
sub_grade                       0.000000
emp_length                      3.784623
home_ownership                  0.000000
annual_inc                      0.003907
verification_status             0.000000
issue_d                         0.000000
loan_status                     0.000000
pymnt_plan                      0.000000
purpose                         0.000000
dti                             0.000000
delinq_2yrs                     0.028324
earliest_cr_line                0.028324
inq_last_6mths                  0.028324
mths_since_last_delinq         58.008751
mths_since_last_record         89.121772
open_acc                        0.028324
pub_rec                         0.028324
revol_bal       

There are 7 columns with more 40% of missing values.

In [13]:
more_40_null_values = df.isnull().sum()/df.shape[0]*100>40
df.drop(df.columns[more_40_null_values],axis=1,inplace=True)

In [14]:
df.isnull().sum()/df.shape[0]*100

loan_amnt                     0.000000
funded_amnt                   0.000000
funded_amnt_inv               0.000000
term                          0.000000
int_rate                      0.000000
installment                   0.000000
grade                         0.000000
sub_grade                     0.000000
emp_length                    3.784623
home_ownership                0.000000
annual_inc                    0.003907
verification_status           0.000000
issue_d                       0.000000
loan_status                   0.000000
pymnt_plan                    0.000000
purpose                       0.000000
dti                           0.000000
delinq_2yrs                   0.028324
earliest_cr_line              0.028324
inq_last_6mths                0.028324
open_acc                      0.028324
pub_rec                       0.028324
revol_bal                     0.000000
revol_util                    0.125015
total_acc                     0.028324
initial_list_status      

Let's work with columns step by step. First one will be 'emp_length'

In [15]:
df['emp_length'].value_counts(dropna = False)

10+ years    29926
2 years       9687
< 1 year      9319
3 years       8888
5 years       7490
1 year        7015
4 years       6556
6 years       5821
7 years       5514
8 years       4634
NaN           3875
9 years       3663
Name: emp_length, dtype: int64

In [16]:
df['emp_length'] = df['emp_length'].str.replace('years', '')
df['emp_length'] = df['emp_length'].str.replace(' ', '')
df['emp_length'] = df['emp_length'].str.replace('year', '')
df['emp_length'] = df['emp_length'].str.replace('<', '')
df['emp_length'] = df['emp_length'].str.replace('+', '')
df['emp_length'].value_counts(dropna = False)

10     29926
1      16334
2       9687
3       8888
5       7490
4       6556
6       5821
7       5514
8       4634
NaN     3875
9       3663
Name: emp_length, dtype: int64

There are 3875 missing values. I suggest to fill it by 1 year, to decrease probability of loan default.

In [17]:
df['emp_length'].fillna(1, inplace=True)
df['emp_length'] = df['emp_length'].astype('int64')

In [18]:
df['emp_length'].value_counts()

10    29926
1     20209
2      9687
3      8888
5      7490
4      6556
6      5821
7      5514
8      4634
9      3663
Name: emp_length, dtype: int64

There are only 4 missing values, which can be filled by median value

In [19]:
df['annual_inc'].fillna(df['annual_inc'].median(), inplace=True)

Colunms delinq_2yrs, inq_last_6mths, open_acc, pub_rec, acc_now_delinq and total_acc contain 29 missing integer values, which can be also filled by median value

In [20]:
df['delinq_2yrs'].fillna(df['delinq_2yrs'].median(), inplace=True)
df['inq_last_6mths'].fillna(df['inq_last_6mths'].median(), inplace=True)
df['open_acc'].fillna(df['open_acc'].median(), inplace=True)
df['pub_rec'].fillna(df['pub_rec'].median(), inplace=True)
df['total_acc'].fillna(df['total_acc'].median(), inplace=True)
df['acc_now_delinq'].fillna(df['acc_now_delinq'].median(), inplace=True)

Colunm revol_util contains 128 missing integer values, column collections_12_mths_ex_med - 145:

In [21]:
df['revol_util'].fillna(df['revol_util'].median(), inplace=True)
df['collections_12_mths_ex_med'].fillna(df['collections_12_mths_ex_med'].median(), inplace=True)

Columns earliest_cr_line, last_pymnt_d, last_credit_pull_d look like datetime. Missing values could be filled by most popular month (mode):

In [22]:
df['last_pymnt_d']= pd.to_datetime(df['last_pymnt_d'] .fillna(str(df['last_pymnt_d'].mode().values[0])),format='%d/%m/%y').dt.month
df['last_credit_pull_d']= pd.to_datetime(df['last_credit_pull_d'] .fillna(str(df['last_credit_pull_d'].mode().values[0])),format='%d/%m/%y').dt.month
df['earliest_cr_line']= pd.to_datetime(df['earliest_cr_line'] .fillna(str(df['earliest_cr_line'].mode().values[0])),format='%d/%m/%y').dt.month

Column 'issue_d' will be filled as year:

In [23]:
dt_series = pd.to_datetime(df['issue_d'], errors = 'coerce')
df['year'] = dt_series.dt.year
df.drop(['issue_d'], axis = 1, inplace = True)

In [24]:
df.isnull().sum()

loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
emp_length                    0
home_ownership                0
annual_inc                    0
verification_status           0
loan_status                   0
pymnt_plan                    0
purpose                       0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                    0
total_acc                     0
initial_list_status           0
out_prncp                     0
out_prncp_inv                 0
total_pymnt                   0
total_pymnt_inv               0
total_rec_prncp               0
total_rec_int                 0
total_re

### 2.2.2. Working with categorical variables

It seems like columns grade, sub_grade, home_ownership, verification_status, pymnt_plan, purpose, initial_list_status and application_type are categorical variables. 


In [25]:
categorical_vars = ['grade', 'sub_grade', 'home_ownership', 'verification_status', 'pymnt_plan', 
                    'purpose', 'initial_list_status', 'application_type']
df_dummies = pd.get_dummies(df[categorical_vars])
df = pd.concat((df, df_dummies),axis=1)

In [26]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0       5000         5000           4975.0   36 months     10.65       162.87   
1       2500         2500           2500.0   60 months     15.27        59.83   
2       2400         2400           2400.0   36 months     15.96        84.33   
3      10000        10000          10000.0   36 months     13.49       339.31   
4       3000         3000           3000.0   60 months     12.69        67.79   

  grade sub_grade  emp_length home_ownership  ...  purpose_medical  \
0     B        B2          10           RENT  ...                0   
1     C        C4           1           RENT  ...                0   
2     C        C5          10           RENT  ...                0   
3     C        C1          10           RENT  ...                0   
4     B        B5           1           RENT  ...                0   

  purpose_moving purpose_other purpose_renewable_energy  \
0              0             0                        0   
1              0             0                        0   
2              0             0                        0   
3              0             1                        0   
4              0             1                        0   

  purpose_small_business  purpose_vacation  purpose_wedding  \
0                      0                 0                0   
1                      0                 0                0   
2                      1                 0                0   
3                      0                 0                0   
4                      0                 0                0   

   initial_list_status_f  initial_list_status_w  application_type_INDIVIDUAL  
0                      1                      0                            1  
1                      1                      0                            1  
2                      1                      0                            1  
3                      1                      0                            1  
4                      1                      0                            1  

[5 rows x 111 columns]

For now we have to drop original columns:

In [27]:
df.drop(['grade', 'sub_grade', 'home_ownership', 'verification_status', 'pymnt_plan', 
                    'purpose', 'initial_list_status', 'application_type'], axis = 1, inplace = True)

In [28]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0       5000         5000           4975.0   36 months     10.65       162.87   
1       2500         2500           2500.0   60 months     15.27        59.83   
2       2400         2400           2400.0   36 months     15.96        84.33   
3      10000        10000          10000.0   36 months     13.49       339.31   
4       3000         3000           3000.0   60 months     12.69        67.79   

   emp_length  annual_inc  loan_status    dti  ...  purpose_medical  \
0          10     24000.0   Fully Paid  27.65  ...                0   
1           1     30000.0  Charged Off   1.00  ...                0   
2          10     12252.0   Fully Paid   8.72  ...                0   
3          10     49200.0   Fully Paid  20.00  ...                0   
4           1     80000.0      Current  17.94  ...                0   

   purpose_moving  purpose_other  purpose_renewable_energy  \
0               0              0                         0   
1               0              0                         0   
2               0              0                         0   
3               0              1                         0   
4               0              1                         0   

   purpose_small_business  purpose_vacation  purpose_wedding  \
0                       0                 0                0   
1                       0                 0                0   
2                       1                 0                0   
3                       0                 0                0   
4                       0                 0                0   

   initial_list_status_f  initial_list_status_w  application_type_INDIVIDUAL  
0                      1                      0                            1  
1                      1                      0                            1  
2                      1                      0                            1  
3                      1                      0                            1  
4                      1                      0                            1  

[5 rows x 103 columns]

We have also column term which is loan duration. Let's make it as integer

In [29]:
df = df.rename(columns={"term": "term_months"})

In [30]:
df['term_months'] = df['term_months'].str.replace(' months', '')
df['term_months'].value_counts(dropna = True)

 36    74821
 60    27567
Name: term_months, dtype: int64

In [31]:
df['term_months'] = df['term_months'].astype('int64')

### 2.2.3. Working with target variable

In [32]:
df['loan_status'].value_counts()

Fully Paid                                             54314
Current                                                32417
Charged Off                                            11237
Does not meet the credit policy. Status:Fully Paid      1988
Late (31-120 days)                                       969
Does not meet the credit policy. Status:Charged Off      761
In Grace Period                                          457
Late (16-30 days)                                        153
Default                                                   92
Name: loan_status, dtype: int64

Rows with loan_status equal to 'Fully Paid', 'Current' and 'Does not meet the credit policy. Status:Fully Paid' can be labeled as good loans (0), others as bad (1)

In [33]:
df['loan_status'] = df['loan_status'].map({'Fully Paid':0,
                       'Current':0, 
                       'Does not meet the credit policy. Status:Fully Paid':0,
                       'Charged Off':1,
                       'Late (31-120 days)':1,
                       'Does not meet the credit policy. Status:Charged Off':1, 
                       'In Grace Period':1,
                       'Late (16-30 days)':1,
                       'Default':1})

In [34]:
y = df['loan_status']
X = df.drop(['loan_status'], axis = 1)

# 3. Binary classification

In [35]:
X_train, X_test, y_train, y_test = X[:70000], X[70000:], y[:70000], y[70000:]

In [36]:
log_res = LogisticRegression()
log_res.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [37]:
prediction = log_res.predict(X_test)

In [38]:
acc_log_res = accuracy_score(prediction, y_test)

In [39]:
acc_log_res

0.9749907373101149